src/client/api/v1/get_balance.c

```diff
@@ -0,0 +1,161 @@
+    }
+    free(res);
+  }
+}
+
+int deser_balance_info(char const *const j_str, res_balance_t *res) {
+  char const *const key_addr = "address";
+  char const *const key_maxResults = "maxResults";
+  char const *const key_count = "count";
+  char const *const key_balance = "balance";
+  char const *const key_code = "code";
+  char const *const key_message = "message";
+  int ret = 0;
+
+  cJSON *json_obj = cJSON_Parse(j_str);
+  if (json_obj == NULL) {
+    return -1;
+  }
+
+  res_err_t *res_err = deser_error(json_obj);
+  if (res_err) {
+    // got an error response
+    res->is_error = true;
+    res->u.error = res_err;
+    ret = 0;
+    goto end;
+  }
+
+  res->u.output_balance = malloc(sizeof(get_balance_t));
+  if (res->u.output_balance == NULL) {
+    printf("[%s:%d] OOM\n", __func__, __LINE__);
+    ret = -1;
+    goto end;
+  }
+
+  cJSON *data_obj = cJSON_GetObjectItemCaseSensitive(json_obj, key_data);
+  if (data_obj) {
+    // gets address
+    if ((ret = json_get_string(data_obj, key_addr, res->u.output_balance->address,
+                               sizeof(res->u.output_balance->address))) != 0) {
+      printf("[%s:%d]: gets %s failed\n", __func__, __LINE__, key_addr);
+      ret = -1;
+      goto end;
+    }
+
+    // gets max_results
+    if ((ret = json_get_uint16(data_obj, key_maxResults, &res->u.output_balance->max_results)) != 0) {
+      printf("[%s:%d]: gets %s json max_results failed\n", __func__, __LINE__, key_maxResults);
+      ret = -1;
+      goto end;
+    }
+
+    // gets count
+    if ((ret = json_get_uint16(data_obj, key_count, &res->u.output_balance->count)) != 0) {
+      printf("[%s:%d]: gets %s json count failed\n", __func__, __LINE__, key_count);
+      ret = -1;
+      goto end;
+    }
+
+    // gets balance
+    if ((ret = json_get_uint64(data_obj, key_balance, &res->u.output_balance->balance)) != 0) {
+      printf("[%s:%d]: gets %s json balance failed\n", __func__, __LINE__, key_balance);
+      ret = -1;
+      goto end;
+    }
+  }
+
+end:
+  cJSON_Delete(json_obj);
+  return ret;
+}
+
+int get_balance(iota_client_conf_t const *conf, char addr[], res_balance_t *res) {
+  int ret = 0;
+  char const *const cmd_balance = "api/v1/addresses/ed25519/";
+
+  if (addr == NULL || res == NULL || conf == NULL) {
+    printf("[%s:%d]: get_balance failed (null parameter)\n", __func__, __LINE__);
+    return -1;
+  }
+
+  if (strlen(addr) != IOTA_ADDRESS_HEX_BYTES) {
+    printf("[%s:%d]: get_balance failed (invalid addr length)\n", __func__, __LINE__);
+    return -1;
+  }
+
+  // compose restful api command
+  iota_str_t *cmd = iota_str_new(conf->url);
+  if (cmd == NULL) {
+    printf("[%s:%d] OOM\n", __func__, __LINE__);
+    return -1;
+  }
+
+  if (iota_str_append(cmd, cmd_balance)) {
+    printf("[%s:%d]: cmd_balance append failed\n", __func__, __LINE__);
+    return -1;
+  }
+
+  if (iota_str_append(cmd, addr)) {
+    printf("[%s:%d]: addr append failed\n", __func__, __LINE__);
+    return -1;
+  }
```

Comment:
these need to revise as well, due to https://github.com/iotaledger/iota.c/pull/71/commits/7bdfdaa364181c61b7fc61a7729ed30302744ac3

Reply:
could you explain in detail?
should I call `iota_str_destroy(cmd)` before `return`?

Reply:
yes, you can fix memory leak issues by one of follows:
a) call `iota_str_destroy` before return
b) goto done, but you need take care `http_res` as well.

Reply:
ok done.